### DISCUSSION TITLE
PnP Core GraphFirst approach

### DISCUSSION THREAD

**Discussion by kavrat** [pnp/pnpcore#1391] [Q&A] [answered]
I'm considering about using PnP Core instead of PnP Framework because of using MS Graph API. But when I setup my project and tried to get list by id app sends request to `/_api/web/lists?$filter=Id+eq+(...)&top=1` instead of using ms graph API request, even if context.GraphFirst = true.

The same situation with getting by title: `/_api/web/lists/getbytitle('T`, and getting list columns `/_api/Web/Lists(guid'_______')/Fields?$top=100.`

My code:

```
var list = await _context.Web.Lists.GetByTitleAsync(listTitle, p => p.Title, p => p.Id, p=>p.Fields);
var list2 = await _context.Web.Lists.GetByIdAsync(list.Id);   
```

I double-checked the setup of the GraphFirst property. But still don't understand why PnP Core still uses these REST API endpoints instead of MS Graph API. Please explain my this behavior.

**Comment by jansenbe** [comment]
@kavrat : given the big parity gaps in list handling between Graph and REST we dropped some of the Graph support for lists. What you'll seen in the library is that we use graph for isolated areas (e.g. taxonomy) and features which are only available in Graph. For areas where there's both SharePoint REST and Graph APIs we mainly use SharePoint REST. For the few cases where there's no Graph or REST we fall back to CSOM

**Reply by kavrat** [reply]
@jansenbe: But in that case, GraphFirst doesn't make more sense. If list handling in REST API is more efficient then ok. 
And one more question about Rate Limit handling. Do all endpoints support Rate Limits handling?

**Reply by jansenbe** [reply, marked as answer]
@kavrat : You can indeed debate over the term Graph first, when building new features we however try to use Graph where possible, it's just with some core SharePoint features that there's too much parity gaps to successfully use Graph over REST in the SDK as we can't limit the SDK functionality to what's just offered by Graph. If you need that then simply use the Graph SDK.

Regarding Rate Limiting, yes, all endpoints do support that but do know that by default it's not turned on in the SDK. See https://pnp.github.io/pnpcore/using-the-sdk/basics-settings.html, more particularly the RateLimiterMinimumCapacityLeft setting. If you want a more advanced sample on how PnP Core rate limiting can be integrated when also using CSOM check the Microsoft 365 Assessment tool code base (see https://github.com/pnp/pnpassessment/issues)

**Reply by kavrat** [reply]
@jansenbe Thank you for the explanation!
